featuregate/registry.go

```diff
@@ -28,43 +38,66 @@ func NewRegistry() *Registry {
 
 // RegisterOption allows to configure additional information about a Gate during registration.
 type RegisterOption interface {
-	apply(g *Gate)
+	apply(g *Gate) error
 }
 
-type registerOptionFunc func(g *Gate)
+type registerOptionFunc func(g *Gate) error
 
-func (ro registerOptionFunc) apply(g *Gate) {
-	ro(g)
+func (ro registerOptionFunc) apply(g *Gate) error {
+	return ro(g)
 }
 
 // WithRegisterDescription adds description for the Gate.
 func WithRegisterDescription(description string) RegisterOption {
-	return registerOptionFunc(func(g *Gate) {
+	return registerOptionFunc(func(g *Gate) error {
 		g.description = description
+		return nil
 	})
 }
 
 // WithRegisterReferenceURL adds a URL that has all the contextual information about the Gate.
-func WithRegisterReferenceURL(url string) RegisterOption {
-	return registerOptionFunc(func(g *Gate) {
-		g.referenceURL = url
+func WithRegisterReferenceURL(referenceURL string) RegisterOption {
+	return registerOptionFunc(func(g *Gate) error {
+		if _, err := url.Parse(referenceURL); err != nil {
+			return fmt.Errorf("WithRegisterReferenceURL: invalid reference URL %q: %w", referenceURL, err)
+		}
+
+		g.referenceURL = referenceURL
+		return nil
 	})
 }
 
+func validateVersion(v string) error {
+	if _, err := version.NewVersion(v); err != nil {
+		return err
+	}
+	return nil
+}
+
 // WithRegisterFromVersion is used to set the Gate "FromVersion".
 // The "FromVersion" contains the Collector release when a feature is introduced.
 func WithRegisterFromVersion(fromVersion string) RegisterOption {
-	return registerOptionFunc(func(g *Gate) {
+	return registerOptionFunc(func(g *Gate) error {
+		if err := validateVersion(fromVersion); err != nil {
+			return fmt.Errorf("WithRegisterFromVersion: %w", err)
+		}
+
 		g.fromVersion = fromVersion
+		return nil
 	})
 }
 
 // WithRegisterToVersion is used to set the Gate "ToVersion".
 // The "ToVersion", if not empty, contains the last Collector release in which you can still use a feature gate.
 // If the feature stage is either "Deprecated" or "Stable", the "ToVersion" is the Collector release when the feature is removed.
 func WithRegisterToVersion(toVersion string) RegisterOption {
-	return registerOptionFunc(func(g *Gate) {
+	return registerOptionFunc(func(g *Gate) error {
+		if err := validateVersion(toVersion); err != nil {
```

Comment:
Should we store the parsed version? Just a question :)

Reply:
It simplifies the code a bit so I went for it. This now means some normalization happens (e.g. `v0.088.0` gets transformed into `v0.88.0`) but I think that's okay.